source/server/callback_router.h

```diff
@@ -16,44 +18,36 @@ template <typename TReturn, typename... TArgs>
 class CallbackRouter {
  public:
   using Handler = std::function<TReturn(TArgs...)>;
-  using Token = void*;
 
   // Move-only RAII type exposed through register_handler, to auto-unregister_handler when leaving scope.
-  class CallbackRegistration {
+  class CallbackRegistration : public nidevice_grpc::AsyncOperationRegistration {
    public:
-    CallbackRegistration() : token_(0) {}
-    CallbackRegistration(Token token) : token_(token) {}
+    CallbackRegistration(AsyncOperationToken token) : token_(token) {}
     CallbackRegistration(const CallbackRegistration&) = delete;
     CallbackRegistration& operator=(const CallbackRegistration&) = delete;
-    CallbackRegistration(CallbackRegistration&& rhs) : token_(rhs.token_) { rhs.token_ = reinterpret_cast<Token>(0); }
-    CallbackRegistration& operator=(CallbackRegistration&& rhs)
-    {
-      std::swap(token_, rhs.token_);
-      return *this;
-    }
+    CallbackRegistration(CallbackRegistration&& rhs) = delete;
+    CallbackRegistration& operator=(CallbackRegistration&& rhs) = delete;
     ~CallbackRegistration()
     {
-      if (token_) {
-        instance().unregister_handler(token_);
-      }
+      instance().unregister_handler(token_);
     }
 
-    Token token()
+    AsyncOperationToken token() override
     {
       return token_;
     }
 
    private:
-    Token token_;
+    AsyncOperationToken token_;
   };
 
-  static CallbackRegistration register_handler(const Handler& handler)
+  static std::unique_ptr<AsyncOperationRegistration> register_handler(const Handler& handler)
```

Comment:
Note: Returning `unique_ptr` makes all the move stuff above unnecessary ^^^. Just delete `move` instead and allow `unique_ptr` to implement it.